core/CronArchive.php

```diff
@@ -356,206 +275,177 @@ public function init()
+            return;
         }
 
-        do {
-            if ($this->isMaintenanceModeEnabled()) {
-                $this->logger->info("Archiving will stop now because maintenance mode is enabled");
-                return;
-            }
-
-            $idSite = $this->websites->getNextSiteId();
-            $numWebsitesArchived++;
+        $countOfProcesses = $this->getMaxConcurrentApiRequests();
 
-            if (null === $idSite) {
-                break;
-            }
+        // if we skip or can't process an idarchive, we want to ignore it the next time we look for an invalidated
+        // archive. these IDs are stored here (using a list like this serves to keep our SQL simple).
+        $this->idArchivesToExclude = [];
 
-            if ($numWebsitesArchived > $numWebsitesScheduled) {
-                // this is needed because a cron:archive might run for example for 5 hours. Meanwhile 5 other
-                // `cron:archive` have been possibly started... this means meanwhile, within the 5 hours, the
-                // `list of SharedSiteIds` have been potentially emptied and filled again from the beginning.
-                // This means 5 hours later, even though all websites that were originally in the list have been
-                // finished by now, the `cron:archive` will stay active and continue processing because the list of
-                // siteIds to archive was reset by another `cron:archive` command. Potentially some `cron:archive`
-                // will basically never end because by the time the `cron:archive` finishes, the sharedSideIds have
-                // been reset. This can eventually lead to some random concurrency issues when there are like
-                // 40 `core:archive` active at the same time.
-                $this->logger->info("Stopping archiving as the initial list of websites has been processed.");
+        $idSite = null;
+        while (true) {
+            if ($this->isMaintenanceModeEnabled()) {
+                $this->logger->info("Archiving will stop now because maintenance mode is enabled");
                 return;
             }
 
             if (!Process::isMethodDisabled('getmypid') && !Process::isMethodDisabled('ignore_user_abort')) {
                 // see https://github.com/matomo-org/wp-matomo/issues/163
                 flush();
             }
-            
-            $requestsBefore = $this->requests;
-            if ($idSite <= 0) {
-                continue;
-            }
 
-            $skipWebsiteForced = in_array($idSite, $this->shouldSkipSpecifiedSites);
-            if ($skipWebsiteForced) {
-                $this->logger->info("Skipped website id $idSite, found in --skip-idsites ");
-                $this->skipped++;
-                continue;
+            if (empty($idSite)) {
+                $idSite = $this->getNextIdSiteToArchive();
```

Comment:
@diosmosis do I understand this correctly that we now first finish archiving all invalidated archives of one site  before moving on to the site?

This means maybe even that  eg
* We launch archiver A which starts on idSite 1
* An hour later we launch another archiver B in parallel and the same one might also start with idSite 1 (vs previously it would have started working on idSite 2)

Also if someone should launch multiple archivers at once they'd all focus on idSite 1 first?

Just wanting to make sure to understand things right. It would mean if that idSite 1 always takes a long time or always fails, it would maybe never archive the other sites?